# Assignment On Model Selection

## Load the dataset into the Python environment

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score

In [2]:
df = pd.read_csv(r'C:\Users\Dell\Desktop\Files\Data Science\Assignments\Assignment 6\titanic_dataset .csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

## Exploratory Data Analysis

In [3]:
df.shape

(891, 12)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [5]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [6]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## Pre-processing

### Handling missing values

In [7]:
obj_na_columns = ['Cabin', 'Embarked']
df[obj_na_columns] = df[obj_na_columns].fillna('Unknown')

In [8]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin            0
Embarked         0
dtype: int64

In [9]:
mean_age = df['Age'].mean()
df['Age'].fillna(mean_age, inplace=True)

In [10]:
df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

### Encoding categorical variables

In [11]:
df.nunique(),df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        891 non-null    object 
 11  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


(PassengerId    891
 Survived         2
 Pclass           3
 Name           891
 Sex              2
 Age             89
 SibSp            7
 Parch            7
 Ticket         681
 Fare           248
 Cabin          148
 Embarked         4
 dtype: int64,
 None)

In [12]:
df1 = df.copy()

In [13]:
df1 = pd.get_dummies(df1, columns=['Sex', 'Embarked'])  #One-hot encoding

### Splitting the dataset into train data and test data

In [14]:
x = df1.select_dtypes(exclude='object')
y = df1['Survived']

In [15]:
x.shape,y.shape

((891, 13), (891,))

In [16]:
scalar = StandardScaler()

In [17]:
x = scalar.fit_transform(x)

In [18]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

## Creatimg KNN and SVM models

In [19]:
knn_classifier = KNeighborsClassifier()
knn_classifier.fit(x_train,y_train)

KNeighborsClassifier()

In [20]:
svc_classifier =SVC()
svc_classifier.fit(x_train,y_train)

SVC()

In [21]:
knn_y_pred = knn_classifier.predict(x_test)
accuracy_knn = accuracy_score(y_test, knn_y_pred)

In [22]:
svc_y_pred = svc_classifier.predict(x_test)
accuracy_svc = accuracy_score(y_test, svc_y_pred)

In [23]:
accuracy_knn

0.9832402234636871

In [24]:
accuracy_svc

0.994413407821229

KNN model = 0.9832  
SVC model = 0.9944

__KNN model with accuracy 98.32 percentage and SVC model with accuracy 99.44 percentage is created__

## K Fold Cross Validation

In [25]:
kfold_validator  = KFold(10)

In [26]:
knn_kfold_cv_result = cross_val_score(knn_classifier,x,y,cv = kfold_validator)
knn_kfold_cv_result

array([0.96666667, 0.96629213, 0.97752809, 0.97752809, 1.        ,
       0.97752809, 1.        , 1.        , 1.        , 0.97752809])

In [27]:
svc_kfold_cv_result = cross_val_score(svc_classifier,x,y,cv = kfold_validator)
svc_kfold_cv_result

array([0.98888889, 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        , 1.        ])

In [28]:
print(f"K-Fold Cross Validation Accuracy for KNN Model : {knn_kfold_cv_result.mean():.4f}")
print(f"K-Fold Cross Validation Accuracy for SVC Model : {svc_kfold_cv_result.mean():.4f}")

K-Fold Cross Validation Accuracy for KNN Model : 0.9843
K-Fold Cross Validation Accuracy for SVC Model : 0.9989


__K-Fold Cross Validation Accuracy for KNN Model : 98.43 %  
K-Fold Cross Validation Accuracy for SVC Model : 99.89 %__

## Stratified Cross Validation

In [29]:
skfold_validator = StratifiedKFold(n_splits = 10)

In [30]:
knn_skfold_cv_result = cross_val_score(knn_classifier,x,y,cv = skfold_validator)
knn_skfold_cv_result

array([0.96666667, 0.97752809, 0.96629213, 0.97752809, 1.        ,
       0.97752809, 1.        , 1.        , 1.        , 0.97752809])

In [31]:
svc_skfold_cv_result = cross_val_score(svc_classifier,x,y,cv = skfold_validator)
svc_skfold_cv_result

array([0.98888889, 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        , 1.        ])

In [32]:
print(f"Stratified Cross Validation Accuracy for KNN Model : {knn_skfold_cv_result.mean():.4f}")
print(f"Stratified Cross Validation Accuracy for SVC Model : {svc_skfold_cv_result.mean():.4f}")

Stratified Cross Validation Accuracy for KNN Model : 0.9843
Stratified Cross Validation Accuracy for SVC Model : 0.9989


__Stratified Cross Validation Accuracy for KNN Model : 98.43 %  
Stratified Cross Validation Accuracy for SVC Model : 99.89 %__

### The average accuracy score of models are  
__K-Fold Cross Validation Accuracy for KNN Model : 98.43 %  
K-Fold Cross Validation Accuracy for SVC Model : 99.89 %  
Stratified Cross Validation Accuracy for KNN Model : 98.43 %  
Stratified Cross Validation Accuracy for SVC Model : 99.89 % %__